### PR TITLE
Improve update_chunk for cframes

### PR DESCRIPTION
**Pull request by martaiborra**
Change `frame_update_chunk` so that when updating a chunk by a smaller one, the new one is placed at the same position and not at the end.

### CODE REVIEW

File: blosc/frame.c [hunk 1]
```diff
@@ -2393,6 +2393,31 @@ void* frame_update_chunk(blosc2_frame_s* frame, int nchunk, void* chunk, blosc2_
...
+    // See how big would be the space
+    old_offset = offsets[nchunk];
+    bool needs_free;
+    uint8_t *chunk_old;
+    int err = blosc2_schunk_get_chunk(schunk, nchunk, &chunk_old, &needs_free);
+    if (err < 0) {
+      BLOSC_TRACE_ERROR("%d chunk con not obtenined from schunk.", nchunk);
+    }
+
+    if (chunk_old == 0) {
```
Comment (aleixalcacer) [inline blosc/frame.c:2408]: It is better to use NULL with pointers.

File: blosc/frame.c [hunk 2]
```diff
@@ -2393,6 +2393,31 @@ void* frame_update_chunk(blosc2_frame_s* frame, int nchunk, void* chunk, blosc2_
...
+      cbytes_old = 0;
+    }
+    else {
+      cbytes_old = sw32_(chunk_old + BLOSC2_CHUNK_CBYTES);
```
Comment (aleixalcacer) [inline blosc/frame.c:2412]: Use `from_big` macro instead of `sw32_`.

File: blosc/schunk.c [hunk 3]
```diff
@@ -686,6 +686,9 @@ int blosc2_schunk_update_chunk(blosc2_schunk *schunk, int nchunk, uint8_t *chunk
         if (frame->sframe) {
           schunk->cbytes -= cbytes_old;
         }
+        else if (cbytes_old >= cbytes) {
```
Comment (aleixalcacer) [inline blosc/schunk.c:689]: Maybe it's better to replace this with:
```
else {  // Contiguous frame
  if (cbytes_old >= cbytes) {
    schunk->cbytes -= cbytes;
  }
}
```

In this way, we visually divide the `sframe` and the `cframe` sections. Then, within the `cframe` section, we will already evaluate the condition.

What do you think?

*The `else if` statement is recommended to use in situations like:*

```
if (year < 18) {
  // Do something
} else if (18 >= year && year < 25) {
  // Do something
} else {
  // Do something
}
```

**Review by aleixalcacer — CHANGES_REQUESTED** [review]

File: blosc/frame.c [hunk 4]
```diff
@@ -2393,6 +2393,31 @@ void* frame_update_chunk(blosc2_frame_s* frame, int nchunk, void* chunk, blosc2_
...
+      if (cbytes_old == BLOSC_MAX_OVERHEAD) {
```
Comment (aleixalcacer) [inline blosc/frame.c:2413]: Why?

File: blosc/frame.c [hunk 5]
```diff
@@ -2421,6 +2446,10 @@ void* frame_update_chunk(blosc2_frame_s* frame, int nchunk, void* chunk, blosc2_
       }
   }
 
+  if (!frame->sframe && cbytes_chunk != 0 && cbytes_old >= cbytes_chunk) {
```
Comment (aleixalcacer) [inline blosc/frame.c:2449]: Why `cbytes_chunk != 0`?

File: blosc/frame.c [hunk 6]
```diff
@@ -2393,6 +2393,31 @@ void* frame_update_chunk(blosc2_frame_s* frame, int nchunk, void* chunk, blosc2_
...
+  int32_t cbytes_old;
+  int32_t old_offset;
+  if (!frame->sframe) {
+    // See how big would be the space
+    old_offset = offsets[nchunk];
+    bool needs_free;
+    uint8_t *chunk_old;
+    int err = blosc2_schunk_get_chunk(schunk, nchunk, &chunk_old, &needs_free);
+    if (err < 0) {
+      BLOSC_TRACE_ERROR("%d chunk con not obtenined from schunk.", nchunk);
```
Comment (aleixalcacer) [inline blosc/frame.c:2405]: You must return an error code after tracing an error.

**Comment by aleixalcacer** [comment]
Have you made sure that any test runs this?

File: blosc/frame.c (hunk 6 above)
Comment (martaiborra) [reply blosc/frame.c]: Thanks, fixed in 78ce44badf0b30b2fe0e3f4cb5451e39f26295cd

File: blosc/frame.c (hunk 1 above)
Comment (martaiborra) [reply blosc/frame.c]: Fixed in 9c3bf17a428af246f12b85fe2816bc0dd5177a98.
The `test_update_chunk` runs this in the case of a contiguous frame.

File: blosc/schunk.c (hunk 3 above)
Comment (martaiborra) [reply blosc/schunk.c]: I see, changed in 9c3bf17a428af246f12b85fe2816bc0dd5177a98

**Review by aleixalcacer — APPROVED** [review]

**Pull request merged**